In [1]:
!pip install plotly

 

In [2]:
print("CHATGPT REVIEW ANALYSIS ")


CHATGPT REVIEW ANALYSIS 


In [3]:
!pip install pandas plotly scikit-learn


In [4]:
!python -m pip install --upgrade pip


In [5]:
import pandas as pd
import plotly
import sklearn

print("Pandas version:", pd.__version__)
print("Plotly version:", plotly.__version__)
print("Scikit-learn version:", sklearn.__version__)



Pandas version: 2.2.3
Plotly version: 6.0.0
Scikit-learn version: 1.6.1


In [6]:
import pandas as pd
import plotly.graph_objects as go
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter
import plotly.express as px
import plotly.io as pio
pio.templates.default = "plotly_white"

# load the dataset
df = pd.read_csv("chatgpt_reviews.csv")

df.head()

Review Id  \
0  6fb93778-651a-4ad1-b5ed-67dd0bd35aac   
1  81caeefd-3a28-4601-a898-72897ac906f5   
2  452af49e-1d8b-4b68-b1ac-a94c64cb1dd5   
3  372a4096-ee6a-4b94-b046-cef0b646c965   
4  b0d66a4b-9bde-4b7c-8b11-66ed6ccdd7da   

                                              Review  Ratings  \
0                                               good        5   
1                                               good        5   
2                                           nice app        5   
3                                           nice, ig        5   
4  this is a great app, the bot is so accurate to...        5   

           Review Date  
0  2024-08-23 19:30:05  
1  2024-08-23 19:28:18  
2  2024-08-23 19:22:59  
3  2024-08-23 19:20:50  
4  2024-08-23 19:20:39

In [7]:
# check for null values
df.isnull().sum()

Review Id      0
Review         6
Ratings        0
Review Date    0
dtype: int64

In [8]:
# replacing null values with empty string
df['Review'] = df['Review'].astype(str).fillna('')

In [9]:
#conforming wether null values are filled or not 
df.isnull().sum()

Review Id      0
Review         0
Ratings        0
Review Date    0
dtype: int64

In [20]:
!pip install notebook jupyter


  Using cached PyYAML-6.0.2-cp310-cp310-win_amd64.whl.metadata (2.1 kB)
   ---------------------------------------- 0.0/13.2 MB ? eta -:--:--
   --- ------------------------------------ 1.0/13.2 MB 6.3 MB/s eta 0:00:02
   ------- -------------------------------- 2.6/13.2 MB 7.6 MB/s eta 0:00:02
   ------------- -------------------------- 4.5/13.2 MB 7.5 MB/s eta 0:00:02
   ------------------------- -------------- 8.4/13.2 MB 10.4 MB/s eta 0:00:01
   ------------------------------- -------- 10.5/13.2 MB 10.4 MB/s eta 0:00:01
   ---------------------------------------- 13.2/13.2 MB 11.5 MB/s eta 0:00:00
   ---------------------------------------- 0.0/11.7 MB ? eta -:--:--
   -------------------- ------------------- 6.0/11.7 MB 28.4 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.7 MB 26.7 MB/s eta 0:00:01
   ---------------------------------------- 11.7/11.7 MB 24.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/10.2 MB ? eta -:--:--
   ------------

In [10]:
%pip install textblob


Note: you may need to restart the kernel to use updated packages.


In [ ]:
import pandas as pd
import plotly.graph_objects as go
from textblob import TextBlob

# Load data from CSV
df = pd.read_csv("chatgpt_reviews.csv")

# Ensure 'Review' column exists and convert to string, replacing NaN values
df['Review'] = df['Review'].astype(str).fillna('')

# Function to determine sentiment polarity
def get_sentiment(review):
    if isinstance(review, str):  # Ensure it's a string
        sentiment = TextBlob(review).sentiment.polarity
        if sentiment > 0:
            return 'Positive'
        elif sentiment < 0:
            return 'Negative'
        else:
            return 'Neutral'
    return 'Neutral'  # Default to Neutral for any invalid values

# Apply sentiment analysis
df['Sentiment'] = df['Review'].apply(get_sentiment)

# Count sentiment distribution
sentiment_distribution = df['Sentiment'].value_counts()

# Create bar chart using Plotly
fig = go.Figure(data=[go.Bar(
    x=sentiment_distribution.index,
    y=sentiment_distribution.values,
    marker_color=['green', 'gray', 'red'],
)])

fig.update_layout(
    title='Sentiment Distribution of ChatGPT Reviews',
    xaxis_title='Sentiment',
    yaxis_title='Number of Reviews',
    width=800,
    height=600
)
# import plotly.io as pio
#pio.renderers.default = "notebook"
#pio.renderers.default = "browser" //to open in browser!!
# fig.show()
import plotly.io as pio
pio.renderers.default = "plotly_mimetype"
fig.show()

In [12]:
!pip install --upgrade plotly


In [22]:
# Analyzing What Users Like About ChatGPT(positive reviews)

#import pandas as pd
import plotly.graph_objects as go
from sklearn.feature_extraction.text import CountVectorizer
from collections import Counter
import plotly.express as px
import plotly.io as pio
# filter reviews with positive sentiment
positive_reviews = df[df['Sentiment'] == 'Positive']['Review']

# use CountVectorizer to extract common phrases (n-grams)
vectorizer = CountVectorizer(ngram_range=(2, 3), stop_words='english', max_features=100)
X = vectorizer.fit_transform(positive_reviews)

# sum the counts of each phrase
phrase_counts = X.sum(axis=0)
phrases = vectorizer.get_feature_names_out()
phrase_freq = [(phrases[i], phrase_counts[0, i]) for i in range(len(phrases))]

# sort phrases by frequency
phrase_freq = sorted(phrase_freq, key=lambda x: x[1], reverse=True)

phrase_df = pd.DataFrame(phrase_freq, columns=['Phrase', 'Frequency'])

fig = px.bar(phrase_df,
             x='Frequency',
             y='Phrase',
             orientation='h',
             title='Top Common Phrases in Positive Reviews',
             labels={'Phrase': 'Phrase', 'Frequency': 'Frequency'},
             width=1000,
             height=600)

fig.update_layout(
    xaxis_title='Frequency',
    yaxis_title='Phrase',
    yaxis={'categoryorder':'total ascending'}
)

fig.show()

In [14]:
#Analyzing What Users Don’t Like About ChatGPT(negative reviews)

negative_reviews = df[df['Sentiment'] == 'Negative']['Review']

X_neg = vectorizer.fit_transform(negative_reviews)

phrase_counts_neg = X_neg.sum(axis=0)
phrases_neg = vectorizer.get_feature_names_out()
phrase_freq_neg = [(phrases_neg[i], phrase_counts_neg[0, i]) for i in range(len(phrases_neg))]
phrase_freq_neg = sorted(phrase_freq_neg, key=lambda x: x[1], reverse=True)
phrase_neg_df = pd.DataFrame(phrase_freq_neg, columns=['Phrase', 'Frequency'])
fig = px.bar(phrase_neg_df,
             x='Frequency',
             y='Phrase',
             orientation='h',
             title='Top Common Phrases in Negative Reviews',
             labels={'Phrase': 'Phrase', 'Frequency': 'Frequency'},
             width=1000,
             height=600)

fig.update_layout(
    xaxis_title='Frequency',
    yaxis_title='Phrase',
    yaxis={'categoryorder':'total ascending'}
)
fig.show()

Common Problems Faced by Users in ChatGPT

In [15]:
# grouping similar phrases into broader problem categories
problem_keywords = {
    'Incorrect Answers': ['wrong answer', 'gives wrong', 'incorrect', 'inaccurate', 'wrong'],
    'App Performance': ['slow', 'lag', 'crash', 'bug', 'freeze', 'loading', 'glitch', 'worst app', 'bad app', 'horrible', 'terrible'],
    'User Interface': ['interface', 'UI', 'difficult to use', 'confusing', 'layout'],
    'Features Missing/Not Working': ['feature missing', 'not working', 'missing', 'broken', 'not available'],
    'Quality of Responses': ['bad response', 'useless', 'poor quality', 'irrelevant', 'nonsense']
}

# initialize a dictionary to count problems
problem_counts = {key: 0 for key in problem_keywords.keys()}

# count occurrences of problem-related phrases in negative reviews
for phrase, count in phrase_freq_neg:
    for problem, keywords in problem_keywords.items():
        if any(keyword in phrase for keyword in keywords):
            problem_counts[problem] += count
            break

problem_df = pd.DataFrame(list(problem_counts.items()), columns=['Problem', 'Frequency'])

fig = px.bar(problem_df,
             x='Frequency',
             y='Problem',
             orientation='h', 
             title='Common Problems Faced by Users in ChatGPT',
             labels={'Problem': 'Problem', 'Frequency': 'Frequency'},
             width=1000,
             height=600)

fig.update_layout(
    plot_bgcolor='white',  
    paper_bgcolor='white', 
    xaxis_title='Frequency',
    yaxis_title='Problem',
    yaxis={'categoryorder':'total ascending'}  
)

fig.show()

Analyzing How Reviews Changed Over Time:

To analyze how the reviews have changed over time, I will:

1.Aggregate the reviews by their sentiment (Positive, Neutral, Negative) over time.

2.Visualize the trends in the volume of each sentiment type over time.

In [18]:
df['Review Date'] = pd.to_datetime(df['Review Date'])

# aggregate sentiment counts by date
sentiment_over_time = df.groupby([df['Review Date'].dt.to_period('M'), 'Sentiment']).size().unstack(fill_value=0)

# convert the period back to datetime for plotting
sentiment_over_time.index = sentiment_over_time.index.to_timestamp()

fig = go.Figure()

fig.add_trace(go.Scatter(x=sentiment_over_time.index, y=sentiment_over_time['Positive'],
                         mode='lines', name='Positive', line=dict(color='green')))
fig.add_trace(go.Scatter(x=sentiment_over_time.index, y=sentiment_over_time['Neutral'],
                         mode='lines', name='Neutral', line=dict(color='gray')))
fig.add_trace(go.Scatter(x=sentiment_over_time.index, y=sentiment_over_time['Negative'],
                         mode='lines', name='Negative', line=dict(color='red')))

fig.update_layout(
    title='Sentiment Trends Over Time',
    xaxis_title='Date',
    yaxis_title='Number of Reviews',
    plot_bgcolor='white',  
    paper_bgcolor='white',  
    legend_title_text='Sentiment',
    xaxis=dict(showgrid=True, gridcolor='lightgray'), 
    yaxis=dict(showgrid=True, gridcolor='lightgray')
)

fig.show()

Analyzing How Often Users Promote ChatGPT Using the concept of Net Promoter Score:

**NPS is calculated based on responses to the question**

Promoters: Respondents who provide a score of 9 or 10.
Passives: Respondents who provide a score of 7 or 8.
Detractors: Respondents who provide a score between 0 and 6.
**NPS= %Promoters − %Detractors**

In [17]:
# define the categories based on the ratings
df['NPS Category'] = df['Ratings'].apply(lambda x: 'Promoter' if x == 5 else ('Passive' if x == 4 else 'Detractor'))

# calculate the percentage of each category
nps_counts = df['NPS Category'].value_counts(normalize=True) * 100

# calculate NPS
nps_score = nps_counts.get('Promoter', 0) - nps_counts.get('Detractor', 0)

# display the NPS Score
nps_score

np.float64(64.35313912172705)

The Net Promoter Score (NPS) for ChatGPT, based on the ratings provided in the dataset, is approximately 64.35. It indicates a strong likelihood that users would recommend ChatGPT to others, as a score above 50 is generally considered excellent.